File: src/main/java/frc/sciborgs/scilib/config/MotorConfig.java
package frc.sciborgs.scilib.config;

import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMaxLowLevel;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;

/**
 * MotorConfig is a builder class for standardization of vendor motor
 * controllers.
 * 
 * <p>
 * Example usage for a CANSparkMax differential drive:
 * </p>
 * 
 * <pre><code>
 * var leftMotor = new MotorConfig()
 *  .setNeutralBehavior(NeutralBehavior.BRAKE)
 *  .setCurrentLimit(20);
 * 
 * var rightMotor = new MotorConfig(leftMotor)
 *  .setInverted(true);
 * 
 * int[] leftPorts = {1, 2, 3};
 * int[] rightPorts = {4, 5, 6};
 * 
 * var leftGroup = new MotorControllerGroup(leftMotor.getCanSparkMax(leftPorts));
 * var rightGroup = new MotorControllerGroup(rightMotor.getCanSparkMax(rightPorts));
 * 
 * var driveTrain = new DifferentialDrive(leftGroup, rightGroup);
 * </pre></code>
 */
public final class MotorConfig {

    public enum MotorType {
        BRUSHED(true),
        BRUSHLESS(false);

        private final boolean brushed;

        private MotorType(boolean brushed) {
            this.brushed = brushed;
        }

        public CANSparkMaxLowLevel.MotorType getREV() {
            return brushed ? CANSparkMaxLowLevel.MotorType.kBrushed : CANSparkMaxLowLevel.MotorType.kBrushless;
        }
    }
    
    public enum NeutralBehavior {
        COAST(true),
        BRAKE(false);

        private final boolean coast;

        private NeutralBehavior(boolean coast) {
            this.coast = coast;
        }

        public CANSparkMax.IdleMode getREV() {
            return coast ? CANSparkMax.IdleMode.kCoast : CANSparkMax.IdleMode.kBrake;
        }
    }
    
    private boolean inverted;
    private MotorType motorType;
    private NeutralBehavior neutralBehavior;
    private double openLoopRampRate;
    private int currentLimit;
    private PIDConstants pidConstants;
    // private EncoderConfig eConfig = new EncoderConfig();

    public MotorConfig() {
        setInverted(false);
        setMotorType(MotorType.BRUSHLESS);
        setNeutralBehavior(NeutralBehavior.COAST);
        // setOpenLoopRampRate(openLoopRampRate)
        // setCurrentLimit(currentLimit)
    }

    public MotorConfig(MotorConfig base) {
        this.inverted = base.isInverted();
        this.motorType = base.getMotorType();
        this.neutralBehavior = base.getNeutralBehavior();
        this.openLoopRampRate = base.getOpenLoopRampRate();
        this.currentLimit = base.getCurrentLimit();
        this.pidConstants = base.getPidConstants();
    }

    /**
     * Creates a CANSparkMax based on configured values
     * @param id the motor controller's device id
     * @return a new CANSparkMax object 
     */
    public CANSparkMax getCanSparkMax(int id){
        var motor = new CANSparkMax(id, motorType.getREV());
        motor.setInverted(inverted);
        motor.setIdleMode(neutralBehavior.getREV());
        motor.setOpenLoopRampRate(openLoopRampRate);
        motor.setSmartCurrentLimit(currentLimit);
        if (pidConstants != null) {
            var pid = motor.getPIDController();
            pid.setP(pidConstants.kp());
            pid.setI(pidConstants.ki());
            pid.setD(pidConstants.kd());
        }
        return motor;
    }

    /**
     * Creates a list of CANSparkMax objects based on their configured values
     * One motor controller will be created per id, in order
     * @param ids
     * @return array of CANSparkMax objects
     */
    public CANSparkMax[] getCanSparkMax(int...ids) {
        var res = new CANSparkMax[ids.length];
        for (int i = 0; i < ids.length; i++) {
            res[i] = getCanSparkMax(ids[i]);
        }
        return res;
    }

    public boolean isInverted() {
        return inverted;
    }

    public MotorConfig setInverted(boolean inverted) {
        this.inverted = inverted;
        return this;
    }

    public MotorType getMotorType() {
        return motorType;
    }

    public MotorConfig setMotorType(MotorType motorType) {
        this.motorType = motorType;
        return this;
    }

    public NeutralBehavior getNeutralBehavior() {
        return neutralBehavior;
    }

    public MotorConfig setNeutralBehavior(NeutralBehavior neutralBehavior) {
        this.neutralBehavior = neutralBehavior;
        return this;
    }

    public double getOpenLoopRampRate() {
        return openLoopRampRate;
    }

    public MotorConfig setOpenLoopRampRate(double openLoopRampRate) {
        this.openLoopRampRate = openLoopRampRate;
        return this;
    }

    public int getCurrentLimit() {
        return currentLimit;
    }

    public MotorConfig setCurrentLimit(int currentLimit) {
        this.currentLimit = currentLimit;
        return this;
    }

    public PIDConstants getPidConstants() {
        return pidConstants;
    }

    public MotorConfig setPidConstants(PIDConstants pidConstants) {
        this.pidConstants = pidConstants;
        return this;
    }
    
}

File: src/main/kotlin/frc/sciborgs/scilib/config/MotorConfig.kt
package frc.sciborgs.scilib.config

import com.ctre.phoenix.motorcontrol.NeutralMode
import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX
import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX
import com.revrobotics.CANSparkMax
import com.revrobotics.CANSparkMaxLowLevel.MotorType

/**
 * MotorConfig is a builder class for standardizing vendor motor controllers.
 *
 * ### Example usage for a CANSparkMax differential drive
 *
 * ```kotlin
 * val leftMotor = MotorConfig(neutralBehavior = NeutralBehavior.BRAKE, currentLimit = 80)
 * val rightMotor = leftMotor.copy(inverted = true)
 *
 * val leftPorts = intArrayOf(1, 2, 3)
 * val rightPorts = intArrayOf(4, 5, 6)
 *
 * val leftGroup = MotorControllerGroup(leftMotor.buildCanSparkMax(*leftPorts, motorType = MotorType.kBrushless))
 * val rightGroup = MotorControllerGroup(rightMotor.buildCanSparkMax(*rightPorts, motorType = MotorType.kBrushless))
 * val driveTrain = DifferentialDrive(leftGroup, rightGroup)
 * ```
 *
 * @author Asa Paparo
 */
data class MotorConfig(
    var inverted: Boolean = false,
    var neutralBehavior: NeutralBehavior = NeutralBehavior.COAST,
    var openLoopRampRate: Double = 0.0,
    var currentLimit: Int = 80,
    var pidConstants: PIDConstants? = null,
) {

  /**
   * Creates a CANSparkMax based on configured values
   *
   * @param id the motor controller's device id
   * @param motorType the [MotorType] of the physical motor ***This could break your motor if it is
   * not set correctly***
   * @return a new CANSparkMax object
   */
  fun buildCanSparkMax(id: Int, motorType: MotorType): CANSparkMax {
    val motor = CANSparkMax(id, motorType)
    motor.restoreFactoryDefaults()
    motor.inverted = inverted
    motor.idleMode = neutralBehavior.rev()
    motor.openLoopRampRate = openLoopRampRate
    motor.setSmartCurrentLimit(currentLimit)
    pidConstants?.also {
      val pid = motor.pidController
      pid.p = it.kp
      pid.i = it.ki
      pid.d = it.kd
    }
    motor.burnFlash()

    return motor
  }

  /**
   * Creates an array of CANSparkMax objects based on configured values.
   *
   * One motor controller will be created per id, in order
   *
   * @param ids any number of motor ids
   * @param motorType the [MotorType] of the physical motors ***This could break your motors if it
   * is not set correctly***
   * @return array of CANSparkMax objects
   */
  fun buildCanSparkMax(vararg ids: Int, motorType: MotorType) =
      Array(ids.size) { buildCanSparkMax(it, motorType) }

  /**
   * Creates a TalonSRX based on configured values
   *
   * @param id the motor controller's device id
   * @param pidSlot the pidSlot of the controller
   * @return a new WPI_TalonSRX object
   */
  fun buildTalonSRX(id: Int, pidSlot: Int = 0): WPI_TalonSRX {
    val motor = WPI_TalonSRX(id)
    motor.configFactoryDefault()
    motor.inverted = this.inverted
    motor.setNeutralMode(neutralBehavior.ctre())
    motor.configOpenloopRamp(openLoopRampRate)
    motor.configPeakCurrentLimit(currentLimit)
    motor.enableCurrentLimit(true)
    pidConstants?.also {
      motor.config_kP(pidSlot, it.kp)
      motor.config_kI(pidSlot, it.ki)
      motor.config_kD(pidSlot, it.kd)
    }
    return motor
  }

  /**
   * Creates an array of TalonSRX objects based on configured values
   *
   * One motor controller will be created per id, in order
   *
   * @param ids any number of motor ids
   * @param pidSlot the pidSlot of the controller
   * @return array of WPI_TalonSRX objects
   */
  fun buildTalonSRX(vararg ids: Int, pidSlot: Int = 0) =
      Array(ids.size) { buildTalonSRX(it, pidSlot) }

  /**
   * Creates a TalonFX based on configured values
   *
   * @param id the motor controller's device id
   * @param pidSlot the pidSlot of the controller
   * @return a new WPI_TalonFX object
   */
  fun buildTalonFX(id: Int, pidSlot: Int = 0): WPI_TalonFX {
    val motor = WPI_TalonFX(id)
    motor.configFactoryDefault()
    motor.inverted = inverted
    motor.setNeutralMode(neutralBehavior.ctre())
    motor.configOpenloopRamp(openLoopRampRate)
    pidConstants?.also {
      motor.config_kP(pidSlot, it.kp)
      motor.config_kI(pidSlot, it.ki)
      motor.config_kD(pidSlot, it.kd)
    }
    return motor
  }

  /**
   * Creates an array of TalonFX objects based on configured values
   *
   * One motor controller will be created per id, in order
   *
   * @param ids any number of motor ids
   * @param pidSlot the pidSlot of the controller
   * @return array of WPI_TalonFX objects
   */
  fun buildTalonFX(vararg ids: Int, pidSlot: Int = 0) =
      Array(ids.size) { buildTalonFX(it, pidSlot) }
}

/** Enum to represent a generic neutral behvavior */
enum class NeutralBehavior(private val coast: Boolean) {
  COAST(true),
  BRAKE(false);

  /** Gets the rev compatible neutral mode */
  fun rev() = if (coast) CANSparkMax.IdleMode.kCoast else CANSparkMax.IdleMode.kBrake

  /** Gets the ctre compatible neutral mode */
  fun ctre() = if (coast) NeutralMode.Coast else NeutralMode.Brake
}
